src/Microsoft.Data/DataFrameBinaryOperations.cs

```diff
@@ -0,0 +1,584 @@
+﻿
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Data
+{
+    public partial class DataFrame
+    {
+        #region Binary Operations
+
+        public DataFrame Add<T>(IReadOnlyList<T> values)
+            where T : struct
+        {
+            if (values.Count != ColumnCount)
+            {
+                throw new ArgumentException($"values.Count {values.Count} must match the number of columns in the table", nameof(values));
+            }
+            var newDataFrame = new DataFrame();
+            for (int ii = 0; ii < ColumnCount; ii++)
+            {
+                PrimitiveDataFrameColumn<T> column = _table.Column(ii) as PrimitiveDataFrameColumn<T>;
+                if (column != null)
+                {
+                    var newColumn = column.Clone();
+                    newColumn._columnContainer.Add(values[ii]);
+                    newDataFrame.InsertColumn(ii, newColumn);
+                }
+            }
+            return newDataFrame;
+        }
+
+        public DataFrame Add<T>(T value)
+            where T : struct
+        {
+            var newDataFrame = new DataFrame();
+            for (int ii = 0; ii < ColumnCount; ii++)
+            {
+                PrimitiveDataFrameColumn<T> column = _table.Column(ii) as PrimitiveDataFrameColumn<T>;
+                if (column != null)
```

Comment:
is it correct that it should jsut return silently if this is null (if it's not the expected type)?

Reply:
Yup, this is wrong. It's hard to support binary ops on generic columns at the moment. My next patch deals with this pretty well though